File: yaazh_admin/lib/core/jpeg_jfif.dart
import 'dart:typed_data';

bool isJpegMagic(List<int> bytes) {
  return bytes.length > 3 && bytes[0] == 0xFF && bytes[1] == 0xD8 && bytes[2] == 0xFF;
}

bool isPngMagic(List<int> bytes) {
  return bytes.length > 3 &&
      bytes[0] == 0x89 &&
      bytes[1] == 0x50 &&
      bytes[2] == 0x4E &&
      bytes[3] == 0x47;
}

/// uCrop JPEGs often have EXIF orientation=0 and no JFIF header.
/// Android ImageDecoder (Huawei/Honor) rejects those with "unimplemented".
Uint8List toJfifJpeg(Uint8List input) {
  if (input.length < 4 || input[0] != 0xFF || input[1] != 0xD8) return input;

  final chunks = <int>[0xFF, 0xD8];
  var i = 2;
  var hasJfif = false;

  while (i + 1 < input.length) {
    if (input[i] != 0xFF) {
      chunks.addAll(input.sublist(i));
      break;
    }
    final marker = input[i + 1];
    if (marker == 0xDA) {
      chunks.addAll(input.sublist(i));
      break;
    }
    if (marker == 0xD9) {
      chunks.addAll(input.sublist(i, i + 2));
      break;
    }
    if (marker == 0x00 || marker == 0x01 || (marker >= 0xD0 && marker <= 0xD7)) {
      chunks.addAll(input.sublist(i, i + 2));
      i += 2;
      continue;
    }
    if (i + 3 >= input.length) {
      chunks.addAll(input.sublist(i));
      break;
    }
    final size = (input[i + 2] << 8) | input[i + 3];
    final end = i + 2 + size;
    if (end > input.length || size < 2) {
      chunks.addAll(input.sublist(i));
      break;
    }
    if (marker == 0xE1) {
      i = end;
      continue;
    }
    if (marker == 0xE0) hasJfif = true;
    chunks.addAll(input.sublist(i, end));
    i = end;
  }

  if (!hasJfif) {
    const jfif = <int>[
      0xFF, 0xE0, 0x00, 0x10, 0x4A, 0x46, 0x49, 0x46, 0x00, 0x01, 0x01, 0x00,
      0x00, 0x01, 0x00, 0x01, 0x00, 0x00,
    ];
    chunks.insertAll(2, jfif);
  }

  return Uint8List.fromList(chunks);
}

File: backend/src/utils/jpeg-jfif.ts
/**
 * uCrop / Android often writes JPEG with EXIF only (no JFIF) and orientation=0.
 * Huawei/Honor ImageDecoder then fails with "unimplemented".
 * Strip APP1 EXIF and ensure a JFIF APP0 so the bytes are a baseline JPEG.
 */
export function toJfifJpeg(input: Buffer): Buffer {
  if (input.length < 4 || input[0] !== 0xff || input[1] !== 0xd8) return input;

  const chunks: Buffer[] = [Buffer.from([0xff, 0xd8])];
  let i = 2;
  let hasJfif = false;

  while (i + 1 < input.length) {
    if (input[i] !== 0xff) {
      chunks.push(input.subarray(i));
      break;
    }
    const marker = input[i + 1] ?? 0;
    if (marker === 0xda) {
      chunks.push(input.subarray(i));
      break;
    }
    if (marker === 0xd9) {
      chunks.push(input.subarray(i, i + 2));
      break;
    }
    if (marker === 0x00 || marker === 0x01 || (marker >= 0xd0 && marker <= 0xd7)) {
      chunks.push(input.subarray(i, i + 2));
      i += 2;
      continue;
    }
    if (i + 3 >= input.length) {
      chunks.push(input.subarray(i));
      break;
    }
    const size = ((input[i + 2] ?? 0) << 8) | (input[i + 3] ?? 0);
    const end = i + 2 + size;
    if (end > input.length || size < 2) {
      chunks.push(input.subarray(i));
      break;
    }
    if (marker === 0xe1) {
      i = end;
      continue;
    }
    if (marker === 0xe0) hasJfif = true;
    chunks.push(input.subarray(i, end));
    i = end;
  }

  if (!hasJfif) {
    const jfif = Buffer.from([
      0xff, 0xe0, 0x00, 0x10, 0x4a, 0x46, 0x49, 0x46, 0x00, 0x01, 0x01, 0x00, 0x00, 0x01, 0x00, 0x01,
      0x00, 0x00,
    ]);
    chunks.splice(1, 0, jfif);
  }

  return Buffer.concat(chunks);
}

export function prepareStoredImage(bytes: Buffer): Buffer {
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    return toJfifJpeg(bytes);
  }
  return bytes;
}
